src/gss_names.c

```diff
@@ -281,6 +281,84 @@ uint32_t gssntlm_import_name(uint32_t *minor_status,
                                        output_name);
 }
 
+size_t gssntlm_get_attrs_count(const struct gssntlm_name_attribute *attrs)
+{
+    size_t attrs_count = 0;
+    if (!attrs) {
+        return 0;
+    }
+    while (attrs[attrs_count].attr_name != NULL) {
+        attrs_count++;
+    }
+
+    return attrs_count;
+}
+
+int gssntlm_copy_attrs(const struct gssntlm_name_attribute *src,
+                       struct gssntlm_name_attribute **dst)
+{
+    struct gssntlm_name_attribute *copied_attrs;
+    size_t attrs_count = gssntlm_get_attrs_count(src);
+
+    if (attrs_count == 0) {
+        *dst = NULL;
+        return 0;
+    }
+
+    copied_attrs = calloc((attrs_count + 1), /* +1 for terminator entry */
+                          sizeof(struct gssntlm_name_attribute));
+    if (copied_attrs == NULL) {
+        return ENOMEM;
+    }
+
+    for (size_t i = 0; i < attrs_count; i++) {
+        /* read-only persistent data ptr can be just copied */
+        copied_attrs[i].attr_name = src[i].attr_name;
+        copied_attrs[i].attr_value.length = src[i].attr_value.length;
+        copied_attrs[i].attr_value.value = malloc(src[i].attr_value.length);
+        if (copied_attrs[i].attr_value.value == NULL) {
+            gssntlm_release_attrs(&copied_attrs);
+            return ENOMEM;
+        }
+        memcpy(copied_attrs[i].attr_value.value, src[i].attr_value.value,
+               src[i].attr_value.length);
+    }
+    /* terminator entry is filled with zeroes by calloc */
+
+    *dst = copied_attrs;
+    return 0;
+}
+
+struct gssntlm_name_attribute *
+    gssntlm_find_attr(struct gssntlm_name_attribute *attrs,
+                      const char *attr_name, size_t attr_name_len)
+{
+    if (!attrs) {
+        return NULL;
+    }
+    for (size_t i = 0; attrs[i].attr_name != NULL; i++) {
+        /* We store attr_name as const static zero-terminated string, so
+         * it is always zero-terminated */
+        size_t len = strlen(attrs[i].attr_name);
```

Comment:
This probably doesn't need to be its own variable (it's only used once and fits on the line).

Reply:
ok